### PR TITLE
plan/executor: handle optimizer hint properly for cartesian join

### PR DESCRIPTION
**Pull request by eurekaka**
### What problem does this PR solve? 

Fix https://github.com/pingcap/tidb/issues/9032

### What is changed and how it works?

- allow merge join executor to handle cartesian product;
- raise warning and choose other join implementations if TIDB_INLJ is specified;

### Check List 

Tests 

 - Unit test

Code changes

N/A

Side effects

N/A

Related changes

 - Need to cherry-pick to the release branch

### CODE REVIEW

File: planner/core/exhaust_physical_plans.go [hunk 1]
```diff
@@ -609,12 +609,16 @@ func (p *LogicalJoin) buildFakeEqCondsForIndexJoin(keys, idxCols []*expression.C
...
 func (p *LogicalJoin) tryToGetIndexJoin(prop *property.PhysicalProperty) ([]PhysicalPlan, bool) {
-	if len(p.EqualConditions) == 0 {
-		return nil, false
-	}
 	plans := make([]PhysicalPlan, 0, 2)
 	rightOuter := (p.preferJoinType & preferLeftAsIndexInner) > 0
 	leftOuter := (p.preferJoinType & preferRightAsIndexInner) > 0
+	if len(p.EqualConditions) == 0 {
+		if leftOuter || rightOuter {
+			warning := ErrInternal.GenWithStack("TIDB_INLJ hint inapplicable for cartesian product")
```
Comment (XuHuaiyu) [inline planner/core/exhaust_physical_plans.go]: Will this be better?
TIDB_INLJ hint is inapplicable without equal ON conditional_expr

File: executor/merge_join_test.go [hunk 2]
```diff
@@ -322,6 +321,17 @@ func (s *testSuite1) TestMergeJoin(c *C) {
 	tk.MustExec("create table s(a int, primary key(a))")
 	tk.MustExec("insert into s value(1)")
 	tk.MustQuery("select /*+ TIDB_SMJ(t, s) */ count(*) from t join s on t.a = s.a").Check(testkit.Rows("4"))
+
+	// Test TIDB_SMJ for cartesian product.
+	tk.MustExec("drop table if exists t")
+	tk.MustExec("create table t(a int)")
+	tk.MustExec("insert into t value(1),(2)")
+	tk.MustQuery("select /*+ TIDB_SMJ(t1, t2) */ * from t t1 join t t2 order by t1.a, t2.a").Check(testkit.Rows(
```
Comment (XuHuaiyu) [inline executor/merge_join_test.go]: Add a test for explain?

**Comment by eurekaka** [comment]
Comments addressed, @XuHuaiyu PTAL

**Pull request merged**